// blackcat.js/dist/modules/module/commandBuilder.js
/**
 * Đại diện cho một bộ xây dựng (builder) cho các lệnh bot.
 * @class
 * @param {Object} options - Các tùy chọn để khởi tạo lệnh bot.
 * @param {boolean} options.owner - Xác định xem lệnh có phải chỉ dành cho chủ bot hay không.
 * @param {number} options.cooldown - Thời gian chờ giữa các lần sử dụng lệnh, mặc định là 3000 ms.
 * @param {Array} options.permissions - Quyền mặc định cho thành viên để chạy lệnh.
 * @param {string} options.description - Mô tả của lệnh.
 * @param {string} options.category - Thư mục chứa lệnh.
 * @param {Array} options.aliases - Danh sách các tên lệnh phụ (aliases).
 * @param {function} options.command - Hàm thực hiện lệnh.
 * @param {string} options.usage - Cách sử dụng lệnh.
 * @param {string} options.name - Tên của lệnh.
 * @returns {commandBuilder} Một phiên bản mới của commandBuilder.
 */

export class commandBuilder {
  constructor(options = {}) {
    const { owner, cooldown = 3000, permissions = [], description, aliases = [], name, usage, category, executeCommand } = options;
    this.cooldown = Number(cooldown);
    this.owner = Boolean(owner);
    this.permissions = permissions;
    this.description = description;
    this.category = category;
    this.aliases = aliases;
    this.usage = usage;
    this.name = name;
    if (typeof executeCommand === "function") {
      this.executeCommand = executeCommand;
    } else return;
  };
  /**
   * Chuyển đổi bộ xây dựng lệnh slash thành đối tượng JSON.
   * @returns {Object} Biểu diễn JSON của bộ xây dựng lệnh slash.
   * @example
   * // Chuyển đổi thành JSON
   * const builder = new CommandBuilder({ name: 'example', description: 'Một lệnh ví dụ', run: message => message.reply('Command executed!') });
   * const commandJSON = builder.toJSON();
   */
  toJSON() {
    return {...this};
  }
};
